src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py

```diff
@@ -355,50 +355,71 @@ def cleanRSE(self, rse):
                     except gfal2.GError as gfalExc:
                         if gfalExc.code == errno.ENOENT:
-                            self.logger.warning("MISSING directory: %s", dirPfn)
+                            self.logger.warning("MISSING directory: %s", dirLfn)
                             rmdirSuccess = True
                         else:
-                            self.logger.error("FAILED to remove directory: %s: GFAL exception: %s", dirPfn, str(gfalExc))
+                            self.logger.error("FAILED to remove nonempty directory: %s: GFAL exception: %s", dirLfn, str(gfalExc))
 
-                    # Then if unable to delete the base directory due to nonEmpty err or similar, try with _purgeTree() recursively
                     if not rmdirSuccess:
-                        self.logger.info("Trying to recursively purge directory: %s:\n", dirPfn)
-                        purgeSuccess = self._purgeTree(ctx, dirPfn)
+                        # If the above operation fails try to execute the directory contents deletion in bulk - full list of files per directory
+                        self.logger.info("Trying to clean the contents of nonempty directory: %s  in file slices of: %s files", dirLfn, self.msConfig["filesToDeleteSliceSize"])
+                        for pfnSlice in list(grouper(pfnList, self.msConfig["filesToDeleteSliceSize"])):
+                            try:
+                                delResult = ctx.unlink(pfnSlice)
+                                # Count all the successfully deleted files (if a deletion was
+                                # successful a value of None is put in the delResult list):
+                                self.logger.debug("RSE: %s, Dir: %s, delResult: %s",
+                                                  rse['name'], dirLfn, pformat(delResult))
+                                for gfalErr in delResult:
+                                    if gfalErr is None:
+                                        filesDeletedSuccess += 1
+                                    else:
+                                        errMessage = os.strerror(gfalErr.code)
+                                        rse['counters']['gfalErrors'].setdefault(errMessage, 0)
+                                        rse['counters']['gfalErrors'][errMessage] += 1
+                            except Exception as ex:
+                                msg = "Error while cleaning RSE: %s. "
+                                msg += "Will retry in the next cycle. Err: %s"
+                                self.logger.exception(msg, rse['name'], str(ex))
+
+                        self.logger.info("RSE: %s, Dir: %s, filesDeletedSuccess: %s",
+                                          rse['name'], dirLfn, filesDeletedSuccess)
+
+                        # Now delete the whole branch, which was previously cleaned file by file
+                        # First try to delete the base directory:
+                        try:
```

Comment:
Given that this same code appears in two different places in this method, I would suggest to create a separate function/method for this.

Reply:
Well this will be tricky... 
There is one slight but important difference: 
In the case of deleting a non-missing and non-empty directory, we also update the  counter `filesDeletedSuccess = len(pfnList)`. Unfortunately it makes difference when we try to delete an already missing directory vs deleting an existing one. When trying to delete a missing directory we  announce the deletion as successful, because the directory is indeed missing, but we do not update the file counters, because in the logic we follow later for announcing the RSE as `clean` is the comparison of files we consider suitable for deletion against those we managed to actually delete, rather than the sum of those which we managed to delete + the already missing ones.

Let me see if I can separate this logic.....

Reply:
Ok, I may be wrong. It seems we are using the Directory counters for announcing the RSE clean, while the file counters are used only for the service reports. Which is indeed logical, and makes things easier ....

Reply:
done